experimental/rocm/rocm_device.c

```diff
@@ -0,0 +1,280 @@
+}
+
+static iree_status_t iree_hal_rocm_device_create_executable_cache(iree_hal_device_t *base_device,
+                                                                  iree_string_view_t identifier,
+                                                                  iree_hal_executable_cache_t **out_executable_cache) {
+  iree_hal_rocm_device_t *device = iree_hal_rocm_device_cast(base_device);
+  return iree_hal_rocm_nop_executable_cache_create(&device->context_wrapper, identifier, out_executable_cache);
+}
+
+static iree_status_t iree_hal_rocm_device_create_executable_layout(
+  iree_hal_device_t *base_device, iree_host_size_t push_constants, iree_host_size_t set_layout_count,
+  iree_hal_descriptor_set_layout_t **set_layouts, iree_hal_executable_layout_t **out_executable_layout) {
+  iree_hal_rocm_device_t *device = iree_hal_rocm_device_cast(base_device);
+  return iree_hal_rocm_executable_layout_create(&device->context_wrapper, set_layout_count, set_layouts, push_constants,
+                                                out_executable_layout);
+}
+
+static iree_status_t iree_hal_rocm_device_create_semaphore(iree_hal_device_t *base_device, uint64_t initial_value,
+                                                           iree_hal_semaphore_t **out_semaphore) {
+  iree_hal_rocm_device_t *device = iree_hal_rocm_device_cast(base_device);
+  return iree_hal_rocm_semaphore_create(&device->context_wrapper, initial_value, out_semaphore);
+}
+
+static iree_status_t iree_hal_rocm_device_queue_submit(iree_hal_device_t *base_device,
+                                                       iree_hal_command_category_t command_categories,
+                                                       iree_hal_queue_affinity_t queue_affinity,
+                                                       iree_host_size_t batch_count,
+                                                       const iree_hal_submission_batch_t *batches) {
+  iree_hal_rocm_device_t *device = iree_hal_rocm_device_cast(base_device);
+  for (int i = 0; i < batch_count; i++) {
+    for (int j = 0; j < batches[i].command_buffer_count; j++) {
+      kernelArrayType *exec = iree_hal_rocm_array_command_buffer_exec(batches[i].command_buffers[j]);
+      for (int k = 0; k < exec->num_of_kernels; k++) {
+        if(exec->kernels[0].func != NULL) printf("This function is NOT dud!\n");
+        printf("grid X:%d\n", exec->kernels[0].gridDimX);
+        printf("grid Y:%d\n", exec->kernels[0].gridDimY);
+        printf("grid Z:%d\n", exec->kernels[0].gridDimZ);
+        printf("block X:%d\n", exec->kernels[0].blockDimX);
+        printf("block Y:%d\n", exec->kernels[0].blockDimY);
+        printf("block Z:%d\n", exec->kernels[0].blockDimZ);
+        printf("Size of kernel Params:%lu\n", sizeof(exec->kernels[0].kernelParams));
+        ROCM_RETURN_IF_ERROR(
+          device->context_wrapper.syms,
+          hipModuleLaunchKernel(exec->kernels[0].func, exec->kernels[0].gridDimX, exec->kernels[0].gridDimY,
```

Comment:
Should this happen at dispatch time instead?

Reply:
done